### PR TITLE
JBTM-3749 Recommend that service writers explicitly set the LRA conte…

### PR DESCRIPTION
**Pull request by mmusgrov**
…xt on outgoing requests
https://issues.redhat.com/browse/JBTM-3749

### CODE REVIEW

File: project/en-US/lra/examples.xml [hunk 1]
```diff
@@ -88,4 +88,37 @@ public Response compensateWork(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId)
...
+<section>
+    <title>Making JAX-RS Invocations from JAX-RS Resource Methods</title>
+    <para>
+        According to the specification the LRA context (<ulink url="https://download.eclipse.org/microprofile/microprofile-lra-1.0-M1/microprofile-lra-spec.html#lra-context"></ulink>) is implicitly propagated to outgoing requests. The Narayana implementation enforces this requirement by storing the LRA context with the thread used to service the incoming request and then reading it if the service methods makes outgoing requests and adds the context to the outgoing headers. Although exhaustive testing of the implementation indicates that this requirement is met there is still a concern because the JAX-RS specification does not mandate that the same thread is used for the incoming and outgoing requests. We therefore recommend that the service writer explicitly set the context on outgoing requests as shown in the following code snippet:
+    </para>
+
+    <programlisting language="java" linenumbering="unnumbered">
+    @LRA(value = LRA.Type.REQUIRED)
+    @PUT
+    public Response addContextHeader(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId) {
```
Comment (jmfinelli) [inline project/en-US/lra/examples.xml:100]: Shouldn't this be `lraURI`?

File: project/en-US/lra/examples.xml [hunk 2]
```diff
@@ -88,4 +88,37 @@ public Response compensateWork(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId)
...
+        if (lraURI != null) {
+            // explicitly set the context
+            builder.header(LRA_HTTP_CONTEXT_HEADER, lraURI);
+        }
+
+        Response response = builder.put(Entity.text("bodyText"));
+            
+        // more business logic
+
+        return Response.ok().entity(lraId.toASCIIString()).build();
```
Comment (jmfinelli) [inline project/en-US/lra/examples.xml:119]: Shouldn't this be `lraURI`?

File: project/en-US/lra/examples.xml (hunk 1 above)
Comment (mmusgrov) [reply project/en-US/lra/examples.xml:100]: I changed it to use lraId throughout.

File: project/en-US/lra/examples.xml (hunk 2 above)
Comment (mmusgrov) [reply project/en-US/lra/examples.xml:117]: I changed it to use lraId throughout.

File: project/en-US/lra/examples.xml [hunk 3]
```diff
@@ -88,4 +88,37 @@ public Response compensateWork(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId)
 }
     </programlisting>
 </section>
+<section>
+    <title>Making JAX-RS Invocations from JAX-RS Resource Methods</title>
+    <para>
+        According to the specification the LRA context (<ulink url="https://download.eclipse.org/microprofile/microprofile-lra-1.0-M1/microprofile-lra-spec.html#lra-context"></ulink>) is implicitly propagated to outgoing requests. The Narayana implementation enforces this requirement by storing the LRA context with the thread used to service the incoming request and then reading it if the service methods makes outgoing requests and adds the context to the outgoing headers. Although exhaustive testing of the implementation indicates that this requirement is met there is still a concern because the JAX-RS specification does not mandate that the same thread is used for the incoming and outgoing requests. We therefore recommend that the service writer explicitly set the context on outgoing requests as shown in the following code snippet:
```
[the remark this answers is not in the document]
Comment (mmusgrov) [reply project/en-US/lra/examples.xml]: That doesn't exist yet so I used `https://download.eclipse.org/microprofile/microprofile-lra-2.0-RC1/microprofile-lra-spec-2.0-RC1.html#_setting_the_context_on_outgoing_jax_rs_requests`

[the remark this answers is not in the document]
Comment (xstefank) [reply project/en-US/lra/examples.xml]: It will when the review finishes but I'm ok with this too. It's the same.

**Review by xstefank — APPROVED** [review]

**Review by jmfinelli — APPROVED** [review]

File: project/en-US/lra/examples.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (mmusgrov) [reply project/en-US/lra/examples.xml]: Ok thanks, I will try to remember to update it when version 2.0 becomes available.

**Pull request merged**